### PR TITLE
Base component tests

### PR DESCRIPTION
**Pull request by njunge94**
## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] I have read the **CONTRIBUTING.md** document.
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

This PR adds tests to the following base components of ZenML:

- BasePipeline
- BaseBackend
- BaseStep
- BaseDatasource
- Repository
- ZenMLConfig
- MetadataWrapper

In addition, e6c0ced fixes a bug in `BasePipeline.get_name_from_pipeline_name`, which would return an incorrect partial name for any pipeline name containing at least one underscore.

Some tests still fail because of errors (e.g. type inconsistencies in `BaseDatasource.sample_data()`) or unexpected effects of some ZenML internals (e.g. ZenMLConfig's `from_config` method). Some fine-tuning there is definitely in order before we can plug these into a CI pipeline e.g. with Github Actions.

Special attention should be directed to the correctness of the `equal_*`fixtures defined in `conftest.py`, which are functions designed to test the equality of two ZenML objects. Since these are relied on heavily by the tests, any bug there could lead to unintended test results and, ultimately, silent errors breaking pipeline reproducibility for pipelines loaded from config objects.

### CODE REVIEW

File: zenml/conftest.py [hunk 1]
```diff
@@ -0,0 +1,241 @@
...
+def cleanup(cleanup_metadata_store, cleanup_artifacts):
+    cleanup_metadata_store()
+    cleanup_artifacts()
+
+
+@pytest.fixture
+def run_test_pipelines():
+    def wrapper():
+        repo: Repository = Repository.get_instance()
+        repo.zenml_config.set_pipelines_dir(pipeline_root)
```
Comment (hamzamaiot) [inline zenml/conftest.py:90]: Why is this neccessary? :-)

File: zenml/conftest.py [hunk 2]
```diff
@@ -0,0 +1,241 @@
...
+    def wrapper(ds1: BaseDatasource, ds2: BaseDatasource, loaded=True):
+        # There can be a "None" datasource in a pipeline
+        if ds1 is None and ds2 is None:
+            return True
+        if sum(d is None for d in [ds1, ds2]) == 1:
+            return False
+
+        equal = False
+        equal |= ds1.name == ds2.name
+        equal |= ds1.schema == ds2.schema
```
Comment (hamzamaiot) [inline zenml/conftest.py:172]: Will this work, as these are both dicts

File: zenml/core/metadata/metadata_wrapper_test.py [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+    with pytest.raises(expected_query_error):
+        _ = mds1.store
+
+
+def test_to_from_config(equal_md_stores):
+    mds1 = ZenMLMetadataStore()
+
+    mds2 = ZenMLMetadataStore.from_config(mds1.to_config())
+
+    # TODO: This fails because from_config throws (base store is
```
Comment (hamzamaiot) [inline zenml/core/metadata/metadata_wrapper_test.py:48]: With the recent refactor, there  are no factories. So lets discuss this TODO

File: zenml/conftest.py [hunk 4]
```diff
@@ -0,0 +1,241 @@
...
+
+@pytest.fixture
+def delete_config():
+    def wrapper(filename):
+        repo: Repository = Repository.get_instance()
+        repo.zenml_config.set_pipelines_dir(pipeline_root)
+
+        try:
+            cfg = os.path.join(pipeline_root, filename)
+            os.remove(cfg)
```
Comment (hamzamaiot) [inline zenml/conftest.py:107]: We can use yaml_utils for this

File: zenml/core/pipelines/base_pipeline_test.py [hunk 5]
```diff
@@ -0,0 +1,217 @@
...
+def test_get_pipeline_config():
+    # test just the pipeline config block added to the steps_config
+    name = "my_pipeline"
+    p: BasePipeline = BasePipeline(name=name)
+
+    config = p.get_pipeline_config()
+
+    p_name = p.pipeline_name
+
+    assert config[keys.PipelineKeys.NAME] == p_name
```
Comment (hamzamaiot) [inline zenml/core/pipelines/base_pipeline_test.py:117]: These keys might change a bit with the `newpipelines` branch but we can refactor in there

**Review by hamzamaiot — COMMENTED** [review]
This is a very through PR, and all the tests are written well. I just left a few comments :+1: but all all relatively minor. great job!

File: zenml/conftest.py (hunk 2 above)
Comment (njunge94) [reply zenml/conftest.py:172]: It does: 

https://stackoverflow.com/questions/17217225/what-does-the-operator-actually-do-on-a-python-dictionary

https://stackoverflow.com/questions/58961051/python-3-6-equality-of-two-dictionaries-with-same-keys-but-in-different-order

Also, it does so independently of order, so we do not need to implement any more elaborate equality tester.

File: zenml/conftest.py (hunk 4 above)
Comment (njunge94) [reply zenml/conftest.py]: Resolved in 5f52815.

**Pull request merged**